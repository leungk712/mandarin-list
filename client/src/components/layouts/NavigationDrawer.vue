<template>
  <div class="ma-4 pa-4">
    <v-card>
      <v-navigation-drawer
        app
        expand-on-hover
        mini-variant-width="60"
        clipped
      >
        <v-list>
          <v-list-item class="px-2">
            <v-list-item-avatar
              v-if="user && userInfo"
              class="mr-2 white--text"
              :color="user.avatarColor"
              size="48"
              tile
            >
              {{ userInitials }}
            </v-list-item-avatar>
          </v-list-item>

          <v-list-item>
            <v-list-item-content v-if="user && userInfo">
              <v-list-item-title class="title">
                {{ userInfo.firstName }} {{ userInfo.lastName }}
              </v-list-item-title>
              <v-list-item-subtitle>{{ userInfo.email }}</v-list-item-subtitle>
            </v-list-item-content>
          </v-list-item>

          <v-list-item>
            <v-list-item-content>
              <v-list-item-title>China <v-icon small>schedule</v-icon>: {{ chinaTime }}</v-list-item-title>
            </v-list-item-content>
          </v-list-item>
        </v-list>

        <v-divider />

        <v-list nav dense>
          <v-list-item link :to="{ name: 'Dashboard' }">
            <v-list-item-icon>
              <v-icon>dashboard</v-icon>
            </v-list-item-icon>
            <v-list-item-title>Dashboard</v-list-item-title>
          </v-list-item>
          <v-list-item link :to="{ name: 'Profile' }">
            <v-list-item-icon>
              <v-icon>person</v-icon>
            </v-list-item-icon>
            <v-list-item-title>Profile</v-list-item-title>
          </v-list-item>
          <v-list-item link :to="{ name: 'TravelChina' }">
            <v-list-item-icon>
              <v-icon>map</v-icon>
            </v-list-item-icon>
            <v-list-item-title>Travel China</v-list-item-title>
          </v-list-item>
          <v-list-item link :to="{ name: 'Resources' }">
            <v-list-item-icon>
              <v-icon>folder</v-icon>
            </v-list-item-icon>
            <v-list-item-title>Resources</v-list-item-title>
          </v-list-item>
          <v-list-item link :to="{ name: 'Code' }">
            <v-list-item-icon>
              <v-icon>code</v-icon>
            </v-list-item-icon>
            <v-list-item-title>Code</v-list-item-title>
          </v-list-item>
          <v-list-item link @click="handleLogout">
            <v-list-item-icon>
              <v-icon>logout</v-icon>
            </v-list-item-icon>
            <v-list-item-title>Logout</v-list-item-title>
          </v-list-item>
        </v-list>
      </v-navigation-drawer>
    </v-card>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Watch } from "vue-property-decorator";
import { namespace, State } from "vuex-class";
import { UserState } from "@/models";
import router from "@/router";
import UserModule from "@/store/modules/user";
import moment from "moment-timezone";

const user = namespace(UserModule.name);

@Component({
  name: "NavigationDrawer"
})
export default class NavigationDrawer extends Vue {
    // ===== Store ===== //
    @State("user") public user!: UserState;
    @user.Action("logout") public logout!: () => void;

    // ===== Data ===== //

    // ===== Methods ===== //
    public handleLogout(): void {
        this.logout()
        router.push({ name: "Home" }).catch(() => {
          return true;
        });
    }

    // ===== Computed ===== //
    get chinaTime() {
      return moment.utc().tz("Asia/Shanghai").format("MM/DD, h:mm:ss a");
    }
    get userInfo() {
        return this.user.user;
    }
    get userInitials() {
        return this.userInfo ? `${this.userInfo.firstName.charAt(0)}${this.userInfo.lastName.charAt(0)}` : "";
    }
}
</script>
