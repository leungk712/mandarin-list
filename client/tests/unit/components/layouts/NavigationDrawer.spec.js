import { createLocalVue, shallowMount } from "@vue/test-utils";
import NavigationDrawer from "@/components/layouts/NavigationDrawer.vue";
import Vue from "vue";
import Vuex from "vuex";
import Vuetify from "vuetify";
import { mockUser } from "../../mocks/mock-user";
import moment from "moment";

Vue.use(Vuetify);

const localVue = createLocalVue();
localVue.use(Vuex);

const elem = document.createElement("div");
if (document.body) {
    document.body.appendChild(elem);
};

document.body.setAttribute("data-app", true);

const actions = {
    logout: jest.fn()
};

const mockStore = {
    user: {
        state: {
            user: mockUser
        },
        actions,
        namespaced: true
    }
};

describe('NavigationDrawer', () => {
    let wrapper;
    const store = new Vuex.Store({
        modules: mockStore
    });
    const mountFunction = (options, store = mockStore) => {
        return shallowMount(NavigationDrawer, {
           attachTo: elem,
           localVue,
           store,
           vuetify: new Vuetify(),
           ...options
        });
    };
    describe("template", () => {
        it("should match snapshot", () => {
            wrapper = mountFunction({}, store);
            expect(wrapper.vm.$el).toMatchSnapshot();
            wrapper.destroy();
        });
    });
    describe("script", () => {
        it("handleLogout method performs the correct actions", async() => {
            wrapper = mountFunction({}, store);
            await wrapper.vm.handleLogout();
            expect(actions.logout).toHaveBeenCalled();
            wrapper.destroy();
        });
        it("handleChinaTime method performs the correct actions", async() => {
            wrapper = mountFunction({
                data() {
                    return {
                        clock: ""
                    }
                }
            }, store);
            const chinaTime = moment.utc().tz("Asia/Shanghai").format("MM/DD, h:mm:ss a");
            await wrapper.vm.handleChinaTime();
            expect(wrapper.vm.clock).toEqual(chinaTime);
            wrapper.destroy();
        });
        it("userInfo() computed prop returns the correct values", () => {
            wrapper = mountFunction({}, store);
            expect(wrapper.vm.userInfo).toEqual(mockUser);
            wrapper.destroy();
        });
        it("userInitials() computed prop returns the correct value", () => {
            wrapper = mountFunction({}, store);
            expect(wrapper.vm.userInitials).toEqual("TU");
            wrapper.destroy();
        })
    });
});