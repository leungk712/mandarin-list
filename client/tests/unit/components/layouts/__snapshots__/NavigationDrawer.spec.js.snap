// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationDrawer template should match snapshot 1`] = `
<div
  class="ma-4 pa-4"
>
  <v-card-stub
    loaderheight="4"
    tag="div"
  >
    <v-navigation-drawer-stub
      app="true"
      class="navigation-drawer"
      clipped="true"
      data-testid="navigation-drawer"
      expandonhover="true"
      height="100vh"
      minivariantwidth="60"
      mobilebreakpoint="1264"
      src=""
      tag="nav"
      width="256"
    >
      <v-list-stub
        tag="div"
      >
        <v-list-item-stub
          activeclass=""
          class="px-2"
          tag="div"
        >
          <v-list-item-avatar-stub
            class="mr-2 white--text"
            size="48"
            tile="true"
          >
            
            TU
          
          </v-list-item-avatar-stub>
        </v-list-item-stub>
         
        <v-list-item-stub
          activeclass=""
          tag="div"
        >
          <v-list-item-content-stub>
            <v-list-item-title-stub
              class="title"
            >
              
              Test User
            
            </v-list-item-title-stub>
             
            <v-list-item-subtitle-stub>
              testuser@email.com
            </v-list-item-subtitle-stub>
          </v-list-item-content-stub>
        </v-list-item-stub>
         
        <v-list-item-stub
          activeclass=""
          tag="div"
        >
          <v-list-item-content-stub>
            <v-list-item-title-stub>
              China 
              <v-icon-stub
                small=""
              >
                schedule
              </v-icon-stub>
              :
              
            </v-list-item-title-stub>
          </v-list-item-content-stub>
        </v-list-item-stub>
      </v-list-stub>
       
      <v-divider-stub />
       
      <v-list-stub
        dense="true"
        nav="true"
        tag="div"
      >
        <v-list-item-stub
          activeclass=""
          class="nav-drawer-dashboard-link"
          data-testid="nav-drawer-dashboard-link"
          link="true"
          tag="div"
          to="[object Object]"
        >
          <v-list-item-icon-stub>
            <v-icon-stub>
              dashboard
            </v-icon-stub>
          </v-list-item-icon-stub>
           
          <v-list-item-title-stub>
            Dashboard
          </v-list-item-title-stub>
        </v-list-item-stub>
         
        <v-list-item-stub
          activeclass=""
          class="nav-drawer-profile-link"
          data-testid="nav-drawer-profile-link"
          link="true"
          tag="div"
          to="[object Object]"
        >
          <v-list-item-icon-stub>
            <v-icon-stub>
              person
            </v-icon-stub>
          </v-list-item-icon-stub>
           
          <v-list-item-title-stub>
            Profile
          </v-list-item-title-stub>
        </v-list-item-stub>
         
        <v-list-item-stub
          activeclass=""
          class="nav-drawer-travel-link"
          data-testid="nav-drawer-travel-link"
          link="true"
          tag="div"
          to="[object Object]"
        >
          <v-list-item-icon-stub>
            <v-icon-stub>
              map
            </v-icon-stub>
          </v-list-item-icon-stub>
           
          <v-list-item-title-stub>
            Travel China
          </v-list-item-title-stub>
        </v-list-item-stub>
         
        <v-list-item-stub
          activeclass=""
          class="nav-drawer-resources-link"
          data-testid="nav-drawer-resources-link"
          link="true"
          tag="div"
          to="[object Object]"
        >
          <v-list-item-icon-stub>
            <v-icon-stub>
              folder
            </v-icon-stub>
          </v-list-item-icon-stub>
           
          <v-list-item-title-stub>
            Resources
          </v-list-item-title-stub>
        </v-list-item-stub>
         
        <v-list-item-stub
          activeclass=""
          class="nav-drawer-code-link"
          data-testid="nav-drawer-code-link"
          link="true"
          tag="div"
          to="[object Object]"
        >
          <v-list-item-icon-stub>
            <v-icon-stub>
              code
            </v-icon-stub>
          </v-list-item-icon-stub>
           
          <v-list-item-title-stub>
            About
          </v-list-item-title-stub>
        </v-list-item-stub>
         
        <v-list-item-stub
          activeclass=""
          class="nav-drawer-logout-link"
          data-testid="nav-drawer-logout-link"
          link="true"
          tag="div"
        >
          <v-list-item-icon-stub>
            <v-icon-stub>
              logout
            </v-icon-stub>
          </v-list-item-icon-stub>
           
          <v-list-item-title-stub>
            Logout
          </v-list-item-title-stub>
        </v-list-item-stub>
      </v-list-stub>
    </v-navigation-drawer-stub>
  </v-card-stub>
</div>
`;
